kernel/src/table_changes/mod.rs

```diff
@@ -0,0 +1,196 @@
+        } else if !is_cdf_enabled(&end_snapshot) {
+            return Err(Error::table_changes_disabled(end_snapshot.version()));
+        }
+        if start_snapshot.schema() != end_snapshot.schema() {
+            return Err(Error::generic(format!(
+                "Failed to build TableChanges: Start and end version schemas are different. Found start version schema {:?} and end version schema {:?}", start_snapshot.schema(), end_snapshot.schema(),
+            )));
+        }
+
+        let log_root = table_root.join("_delta_log/")?;
+        let log_segment = LogSegment::for_table_changes(
+            engine.get_file_system_client().as_ref(),
+            log_root,
+            start_version,
+            end_version,
+        )?;
+
+        let schema = StructType::new(
+            end_snapshot
+                .schema()
+                .fields()
+                .cloned()
+                .chain(CDF_FIELDS.clone()),
+        );
+
+        Ok(TableChanges {
+            table_root,
+            end_snapshot,
+            log_segment,
+            start_version,
+            schema,
+        })
+    }
+    pub fn start_version(&self) -> Version {
+        self.start_version
+    }
+    /// The end version of the `TableChanges`. If no end_version was specified in
+    /// [`TableChanges::try_new`], this returns the newest version as of the call to `try_new`.
+    pub fn end_version(&self) -> Version {
+        self.log_segment.end_version
+    }
+    /// The logical schema of the change data feed. For details on the shape of the schema, see
+    /// [`TableChanges`].
+    pub fn schema(&self) -> &Schema {
+        &self.schema
+    }
+    pub fn table_root(&self) -> &Url {
+        &self.table_root
+    }
+    #[allow(unused)]
+    pub(crate) fn partition_columns(&self) -> &Vec<String> {
+        &self.end_snapshot.metadata().partition_columns
+    }
+    #[allow(unused)]
+    pub(crate) fn column_mapping_mode(&self) -> &ColumnMappingMode {
+        &self.end_snapshot.column_mapping_mode
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::engine::sync::SyncEngine;
+    use crate::schema::DataType;
+    use crate::schema::StructField;
+    use crate::table_changes::CDF_FIELDS;
+    use crate::Error;
+    use crate::Table;
+    use itertools::assert_equal;
+
+    #[test]
+    fn test_enable_cdf_flag() {
+        let path = "./tests/data/table-with-cdf";
+        let engine = Box::new(SyncEngine::new());
+        let table = Table::try_from_uri(path).unwrap();
+
+        let valid_ranges = [(0, 1), (0, 0), (1, 1)];
+        for (start_version, end_version) in valid_ranges {
+            let table_changes = table
+                .table_changes(engine.as_ref(), start_version, end_version)
+                .unwrap();
+            assert_eq!(table_changes.start_version, start_version);
+            assert_eq!(table_changes.end_version(), end_version);
+        }
+
+        let invalid_ranges = [(0, Some(2)), (1, Some(2)), (2, Some(2))];
```

Comment:
can we do this? also what about (2, 3)?
```suggestion
        let invalid_ranges = [(0, 2), (1, 2), (2, 2), (2, 3)];
```